File: lab-3/python/bai2_3.py
import pandas as pd
import numpy as np
import statsmodels.api as sm
from sklearn.metrics import mean_absolute_error, mean_squared_error

data = pd.read_csv("Home_Market_Value.csv")

X1 = sm.add_constant(data["House Age"])
y = data["Market Value"]

model1 = sm.OLS(y, X1).fit()
y_pred1 = model1.predict(X1)


X2 = sm.add_constant(data[["House Age", "Square Feet"]])
model2 = sm.OLS(y, X2).fit()
y_pred2 = model2.predict(X2)

def regression_metrics(y_true, y_pred, model):
    mae = mean_absolute_error(y_true, y_pred)
    mape = np.mean(np.abs((y_true - y_pred) / y_true)) * 100
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)
    r2 = model.rsquared
    adj_r2 = model.rsquared_adj
    return [mae, mape, mse, rmse, r2, adj_r2]

metrics = pd.DataFrame(
    {
        "Model 1: House Age": regression_metrics(y, y_pred1, model1),
        "Model 2: House Age + Square Feet": regression_metrics(y, y_pred2, model2)
    },
    index=["MAE", "MAPE (%)", "MSE", "RMSE", "R²", "Adj R²"]
)

print("\n Regression Comparison:")
print(metrics)

File: lab-3/r/bai2_3.r
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

data <- read.csv("Home_Market_Value.csv")

model_single <- lm(Market.Value ~ House.Age, data = data)

model_multiple <- lm(Market.Value ~ House.Age + Square.Feet, data = data)


if(!require(Metrics)) install.packages("Metrics")
library(Metrics)

pred_single <- predict(model_single, data)
pred_multiple <- predict(model_multiple, data)

actual <- data$Market.Value

MAE_single  <- mae(actual, pred_single)
MAPE_single <- mean(abs((actual - pred_single) / actual)) * 100
MSE_single  <- mse(actual, pred_single)
RMSE_single <- rmse(actual, pred_single)
R2_single   <- summary(model_single)$r.squared
AdjR2_single <- summary(model_single)$adj.r.squared

MAE_multi  <- mae(actual, pred_multiple)
MAPE_multi <- mean(abs((actual - pred_multiple) / actual)) * 100
MSE_multi  <- mse(actual, pred_multiple)
RMSE_multi <- rmse(actual, pred_multiple)
R2_multi   <- summary(model_multiple)$r.squared
AdjR2_multi <- summary(model_multiple)$adj.r.squared

compare_models <- data.frame(
  Metric = c("MAE", "MAPE (%)", "MSE", "RMSE", "R-squared", "Adjusted R-squared"),
  Single_Model = c(MAE_single, MAPE_single, MSE_single, RMSE_single, R2_single, AdjR2_single),
  Multiple_Model = c(MAE_multi, MAPE_multi, MSE_multi, RMSE_multi, R2_multi, AdjR2_multi)
)

print(compare_models)
